apis/v1alpha2/validation/grpcroute.go

```diff
@@ -53,9 +72,166 @@ func validateGRPCRouteRules(rules []gatewayv1a2.GRPCRouteRule, path *field.Path)
 func validateRuleMatches(matches []gatewayv1a2.GRPCRouteMatch, path *field.Path) field.ErrorList {
 	var errs field.ErrorList
 	for i, m := range matches {
-		if m.Method != nil && m.Method.Service == nil && m.Method.Method == nil {
-			errs = append(errs, field.Required(path.Index(i).Child("methods"), "one or both of `service` or `method` must be specified"))
-			return errs
+		if m.Method != nil {
+			errs = append(errs, validateGRPCMethod(*m.Method, path.Index(i).Child("method"))...)
+		}
+		if m.Headers != nil {
+			errs = append(errs, validateGRPCHeaderMatches(m.Headers, path.Index(i).Child("headers"))...)
+		}
+	}
+	return errs
+}
+
+// validateGRPCMethod validates that no header name
+// is matched more than once (case-insensitive).
+func validateGRPCMethod(m gatewayv1a2.GRPCMethodMatch, path *field.Path) field.ErrorList {
+	var errs field.ErrorList
+	if m.Service == nil && m.Method == nil {
+		return append(errs, field.Required(path, "one or both of `service` or `method` must be specified"))
+	}
+
+	if m.Service != nil {
+		for _, invalidSeq := range invalidPathSequences {
+			if strings.Contains(*m.Service, invalidSeq) {
+				errs = append(errs, field.Invalid(path.Child("service"), *m.Service, fmt.Sprintf("must not contain %q", invalidSeq)))
+			}
+		}
+
+		for _, invalidSuff := range invalidPathSuffixes {
+			if strings.HasSuffix(*m.Service, invalidSuff) {
+				errs = append(errs, field.Invalid(path.Child("service"), *m.Service, fmt.Sprintf("cannot end with '%s'", invalidSuff)))
+			}
+		}
+	}
+
+	if m.Method != nil {
```

Comment:
This is already validated in the CRD:

```

// MethodMatchType specifies the semantics of how gRPC methods and services are compared.
// Valid MethodMatchType values are:
//
// * "Exact"
// * "RegularExpression"
//
// Exact paths MUST be syntactically valid:
//
// - Must not contain `/` character
//
// +kubebuilder:validation:Enum=Exact;PathPrefix;RegularExpression
// +kubebuilder:validation:Enum=Exact;RegularExpression
type GRPCMethodMatchType string
```

(note: it is listed 2x for some reason)

Reply:
I _think_ the first will be immediately overridden by the second, but it's incorrect anyway, so we should just remove it.

Reply:
I can't find that line doubled up, but let me know if it's still in main somewhere that I can't find. @howardjohn which specific part are you saying is validated in the CRD? That m.Method can't be nil?

Reply:
Yeah, I can't find the double up either.